subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+
+    public NormalizedContentInfo collectContentInfo(InputStream inputStream) {
+        FileContentTypeDetectingInputStream contentTypeDetectingInputStream = new FileContentTypeDetectingInputStream(inputStream);
+        ShortCircuitingInputStream<FileContentTypeDetectingInputStream> shortCircuitingInputStream = new ShortCircuitingInputStream<>(
+            contentTypeDetectingInputStream,
+            stream -> stream.getContentType() == FileContentType.BINARY
+        );
+        InputStream normalizedInputStream = new LineEndingNormalizingInputStream(shortCircuitingInputStream);
+
+        try {
+            HashCode hashCode = streamHasher.hash(normalizedInputStream);
+            return new NormalizedContentInfo(hashCode, contentTypeDetectingInputStream.getContentType(), shortCircuitingInputStream.isShortCircuited());
+        } finally {
+            closeAndIgnore(contentTypeDetectingInputStream, shortCircuitingInputStream, normalizedInputStream);
+        }
+    }
+
+    public NormalizedContentInfo collectContentInfo(File file) {
+        FileInputStream inputStream = null;
+        try {
+            inputStream = new FileInputStream(file);
+            return collectContentInfo(inputStream);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(String.format("Failed to create MD5 hash for file '%s' as it does not exist.", file), e);
+        } finally {
+            closeAndIgnore(inputStream);
+        }
+    }
+
+    private static void closeAndIgnore(InputStream... inputStreams) {
+        for (InputStream inputStream : inputStreams) {
+            try {
+                if (inputStream != null) {
+                    inputStream.close();
+                }
+            } catch (IOException e) {
+                // Ignored
+            }
+        }
+    }
+
+    private static class NormalizedContentInfo {
+        private final HashCode hash;
+        private final FileContentType contentType;
+        private final boolean incompleteHash;
+
+        public NormalizedContentInfo(HashCode hash, FileContentType contentType, boolean incompleteHash) {
+            this.hash = hash;
+            this.contentType = contentType;
+            this.incompleteHash = incompleteHash;
+        }
+
+        public FileContentType getContentType() {
+            return contentType;
+        }
+
+        public HashCode getHash() {
+            // If reading of the input stream was short circuited because we detected a binary file,
+            // we should never use the hash as it's incomplete.
+            if (incompleteHash) {
+                throw new IllegalStateException();
+            } else {
+                return hash;
+            }
+        }
+    }
+
+    /**
+     * An InputStream that stops the reading from its delegate if the supplied condition is ever met.
+     */
+    private static class ShortCircuitingInputStream<T extends InputStream> extends FilterInputStream {
```

Comment:
IIUC, by implementing this logic as an `InputStream` we get to use `StreamHasher` and avoid having to re-implement the copying logic. However, we still have to write a bunch of code, and the resulting solution becomes quite complex with the wrapping of streams, the predicate, and tracking the short-circuiting after the fact.

I wonder if we could achieve better results by not using `StreamHasher`, and instead doing the reading manually here, just doing a naive implementation with `if`s and loops. I'd probably also try to fold both `LineEndingNormalizingInputStream` and `FileContentTypeDetectingInputStream` into this class and see how the resulting code looks like.

Reply:
I was skeptical that this would actually improve things.  What I liked about the input streams was that the responsibilities were cleanly divided and easy to test and I thought that lumping them together would make it more complex.  But I went ahead and did it anyways and your instincts were correct.  It's actually not that bad.  Thanks for the suggestion.